Add support for timeouts to Mutes cog

TrustyJAID opened the pull request:
### Description of the changes

Add timeouts support to the Mutes cog. This changes the default server mute behavior to utilize timeouts while maintaining role mutes if they have been previously setup. This adds a new `[p]timeout` command that exclusively uses discords built in timeout feature. This also allows the `[p]activemutes` command to list the currently muted and timed out users in the server. This also adds some models that make it easier for managing typed responses for internal muting methods.

### Have the changes in this PR been tested?

<!--
Choose one (remove the line that doesn't apply):
-->
Yes
<!--
If the question doesn't apply (for example, it's not a code change), choose Yes.

Please respond to this question truthfully. We do not delay nor reject PRs
based on the answer to this question but it allows to better review this PR.
-->

TrustyJAID committed:
Mutes utilize timeouts instead of channel overwrites for server mutes. This removes a lot of unnecessary checks previously handled and allows timeouts to be the default if no role is setup for role based mutes. This also adds a dedicated `[p]timeout` to explicitly use discord timeouts as essentially an alias for `[p]mute` but allows for both methods to exist simultaneously.

TrustyJAID committed:
Don't break everyone using i18n

TrustyJAID committed:
Add dedicated response models for muting and add missing checks for timeouts

TrustyJAID committed:
update for d.py 2.0

TrustyJAID committed:
Fix typo causing flake8 to fail lol

TrustyJAID committed:
remove extra newline

TrustyJAID committed:
Fix auto unmutes and voice channels for new result models

TrustyJAID committed:
Missed black on this file 🤔

TrustyJAID committed:
Use discord.utils.format_dt instead of manually creating the timestamp strings

TrustyJAID committed:
Fix unmute bug

TrustyJAID committed:
Merge branch 'V3/develop' into mutes-add-timeouts

TrustyJAID committed:
Actually update for mutes changes

TrustyJAID committed:
Merge branch 'V3/develop' into mutes-add-timeouts

TrustyJAID committed:
Merge branch 'V3/develop' into mutes-add-timeouts

TrustyJAID committed:
Merge branch 'V3/develop' into mutes-add-timeouts

TrustyJAID committed:
Fix duplicate imports and channelmute/channelunmute errors

TrustyJAID committed:
This should not have changed

BlizzardTheWolf commented:
Could we make it so that you can choose wether to use timeouts or mutes by using like `[p]muteset usetimeouts`, and having the command stay `[p]mute`, instead of having it be `[p]timeout`?

AAA3A-AAA3A commented:
> Could we make it so that you can choose wether to use timeouts or mutes by using like `[p]muteset usetimeouts`, and having the command stay `[p]mute`, instead of having it be `[p]timeout`?

I think we should have a `[p]timeout` command, but a setting could redirect `[p]mute` to `[p]timeout` if server's admins prefer this option to a real role.

BlizzardTheWolf commented:
Yeah that's a good idea!

> > Could we make it so that you can choose wether to use timeouts or mutes by using like `[p]muteset usetimeouts`, and having the command stay `[p]mute`, instead of having it be `[p]timeout`?
> 
> I think we should have a `[p]timeout` command, but a setting could redirect `[p]mute` to `[p]timeout` if server's admins prefer this option to a real role.

Flame442 committed:
Merge branch 'V3/develop' into mutes-add-timeouts

Flame442 commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -902,7 +905,7 @@ async def mute_role(self, ctx: commands.Context, *, role: discord.Role = None):
             await self.config.guild(ctx.guild).sent_instructions.set(False)
             # reset this to warn users next time they may have accidentally
             # removed the mute role
-            await ctx.send(_("Channel overwrites will be used for mutes instead."))
```
If this PR is entirely replacing channel override mutes with timeout mutes, the docstring for this command and the output of `[p]muteset force_role_mutes` needs to be update to reflect that.

Flame442 commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -902,7 +905,7 @@ async def mute_role(self, ctx: commands.Context, *, role: discord.Role = None):
             await self.config.guild(ctx.guild).sent_instructions.set(False)
             # reset this to warn users next time they may have accidentally
             # removed the mute role
```
It seems like this is no longer used, so it should probably be removed from here and the defaults.

Flame442 commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -1467,7 +1483,7 @@ async def unmute(
                     )
                 else:
-                    issue_list.append(success)
+                    issue_list.append(response)
```
This appears to be appending a failing response twice. I believe the above check is because the failure message for moving users is successful with a reason, however it will also append an issue in cases where there was not a success, since they to have a reason set. They would then be appended again by this line.

Flame442 commented on redbot/cogs/mutes/voicemutes.py:
```diff
@@ -222,7 +222,7 @@ async def unmute_voice(
                     )
                 else:
-                    issue_list.append((user, success["reason"]))
+                    issue_list.append((user, result.reason))
```
Same issue with potentially double appending as before.

TrustyJAID replied:
This, and the previous mention of a similar situation, are both intended. There are two lists we're dealing with here one for successful mutes we can largely ignore and another for ones with potential problems.

The only situation where we have successfully muted the user but still have a reason is when the user is currently in a voice channel and the bot was unable to move the user in the voice channel triggering discord to actually mute them. Therefore it should be notified to the command issuer that we were able to mute them but they may still be speaking in the voice channel and further action may need to be taken.

Flame442 replied:
I missed the if/else block...

Flame442 commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -1560,43 +1576,31 @@ async def mute_user(
+
         if permissions.administrator:
-            ret["reason"] = _(MUTE_UNMUTE_ISSUES["is_admin"])
+            ret.reason = _(MUTE_UNMUTE_ISSUES["is_admin"])
```
This isn't strictly an error with your PR, but there is no reason to be translating the strings both here and in the definition of `MUTE_UNMUTE_ISSUES`. This applies to this whole function, `unmute_user`, `channel_mute_user`, and `channel_unmute_user`.

TrustyJAID replied:
Actually this *is* necessary since the call to translate the string does not actually happen until here with the correct context. The initialized variables can be argued for not needing it and I can't remember the exact reason for why we left it this way and I can't find the discussion on it now. I just tested the change and can confirm that removing the translation `_()` from any of these cause the result to not correctly translate the message.

Jackenmen replied:
`_()` calls in globals are so that redgettext can collect these strings but they're just passthrough and don't do anything (nor should they since the translations need to happen in a context where the locale is properly set).

`_()` calls in actual code are so that the strings actually get translated but those won't be collected by redgettext since they don't operate on string literals.

Flame442 reviewed (changes requested):
Various developer-facing documentation should be updated to reflect an accurate description of the behavior under timeouts rather than channel override based mutes (lines 101, 315, 771).

I have not yet run the code, these are the issues I found from simply reviewing it.

Flame442 commented:
@BlizzardTheWolf @AAA3A-AAA3A, the PR currently replaces the backup behavior for `[p]mute` when a mute role is not configured to use timeouts, rather than channel overrides. The `[p]timeout` command also exists for explicitly timing out a user, regardless of whether or not `[p]mute` is configured to use a role.

TrustyJAID committed:
Merge branch 'V3/develop' into mutes-add-timeouts

TrustyJAID committed:
Mention use of timeouts instead of channel overwrites in force role mutes command

TrustyJAID committed:
Only add to the issue list once if we are unable to move the user in a voice channel but otherwise able to mute them.

TrustyJAID committed:
Remove last instances of `sent_instruction` since they're no longer necessary

Flame442 commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -831,7 +833,7 @@ async def force_role_mutes(self, ctx: commands.Context, true_or_false: bool):
             await ctx.send(_("Okay I will enforce role mutes before muting users."))
         else:
-            await ctx.send(_("Okay I will allow channel overwrites for muting users."))
+            await ctx.send(_("Okay I will use timeouts for muting users."))
```
I don't think this makes it use timeouts, it just allows timeouts to be used at all. I'm not sure if this setting is even needed, now that there's no API spam to need to opt in to.

TrustyJAID committed:
Clarify wording on forcing role mutes deactivation

TrustyJAID committed:
Clarify docstrings in mute role setup command about using timeouts instead

TrustyJAID committed:
Remove extra permission decorator

TrustyJAID committed:
Replace checks to force role mutes with checks for timeout permissions

TrustyJAID committed:
Allow the unmute_user method to remove all mute methods on a user. This removes some old error messages attempting to unmute a user and removes the need for a new command to specifically remove a timeout from users when a role is set but a timeout is used instead.

TrustyJAID committed:
Add a specific permission issue string regarding members timed out

TrustyJAID committed:
Include length of the mute and until in `[p]timeout` and `[p]mute` and ensure they're consistent

TrustyJAID committed:
Don't attempt to concatenate a None type and a string

TrustyJAID committed:
Add timeout permission issue in mute_user method

TrustyJAID committed:
Clarify wording on issues to include unmuted

TrustyJAID committed:
Attempt to move a user in their voice channel so when a mute is issued they actually receive a voice mute during server wide mutes not just channel mutes.

TrustyJAID committed:
Send `[p]activemutes` as a menu instead of multiple messages resolves #6263
Update issue confirmation list to use buttons instead of reactions or messages if reactions are not possible.

TrustyJAID committed:
Revert "Send `[p]activemutes` as a menu instead of multiple messages resolves #6263"

This reverts commit 4939180e2aa367742ee4f10694f6a2099193c73f.

Flame442 committed:
Fix comment

TrustyJAID committed:
Include hidden `[p]forceunmute` command to fix pre-existing channel override server-wide mutes

TrustyJAID committed:
Style

TrustyJAID committed:
Remove check for mute role in `[p]forceunmute`

TrustyJAID committed:
Fix logic on when to add to the success list and notify user via DM

Jackenmen commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -1467,7 +1477,78 @@ async def unmute(
+        async with ctx.typing():
+            guild = ctx.guild
+            author = ctx.author
+            issue_list = []
```
Nitpick but `[p]unmute`/`[p]channelunmute` sets `audit_reason` generated from `reason` and passes that to `channel_unmute_user()` instead:
```suggestion
            audit_reason = get_audit_reason(author, reason, shorten=True)
            issue_list = []
```

TrustyJAID committed:
Use the same audit reason as other unmute methods in `[p]forceunmute`

TrustyJAID committed:
Improve issue list display by grouping by unique users and include the channel mention on issues with channel overwrite mutes

TrustyJAID committed:
excaped

TrustyJAID committed:
Improve verbosity of some mute issue strings by including a location variable

Jackenmen commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -1411,8 +1428,8 @@ async def channel_mute(
-            for user, issue in issue_list:
-                msg += f"{user}: {issue}\n"
+            for issue in issue_list:
+                msg += f"{issue.user}: {issue.reason}\n"
```
```suggestion
                msg += f"- {issue.user}: {issue.reason}\n"
```

Jackenmen commented on redbot/cogs/mutes/mutes.py:
```diff
@@ -1549,8 +1643,8 @@ async def unmute_channel(
-            for user, issue in issue_list:
-                msg += f"{user}: {issue}\n"
+            for issue in issue_list:
+                msg += f"{issue.user}: {issue.reason}\n"
```
```suggestion
                msg += f"- {issue.user}: {issue.reason}\n"
```

TrustyJAID committed:
Various formatting consistency changes and bullet point issues in mute/unmute server responses

Jackenmen reviewed (approved):
![meow?](https://i.fluffy.cc/hc8WTNRNffFv0fTPJDTdFFbsKMCQn5Rz.gif)
